text/3742-mutate-non-mutable-locals.md

```diff
@@ -0,0 +1,249 @@
+- Feature Name: `mut_non_mut`
+- Start Date: 2024-12-13
+- RFC PR: [rust-lang/rfcs#3742](https://github.com/rust-lang/rfcs/pull/3742)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+the `mut` keyword has two meanings currently:
+ - When applied to a reference, it means the reference is not aliased
+ - When applied to a local binding, it means that the binding is allowed be re-assigned or borrowed using `&mut`
+
+The first of these is fundamental to Rust's safety guarantees and enables some optimizations.  The second,
+however, is more of a lint: the error message may help catch mistakes, but ignoring it doesn't violate
+type safety or subvert any of Rust's global correctness guarantees.
+
+This RFC proposes that the re-assignment or creation of a `&mut` to a local variable not
+marked with `mut` be changed from a hard error to a deny-by-default lint diagnostic (called
+`mut_non_mut` for the sake of discussion).
```

Comment:
What's an error and what's a lint is a mixture of implementation detail and user configuration. Abstracting those details away, what this RFC proposes is relaxing a forbid-by-default diagnostic to deny-by-default. 

```suggestion
This RFC proposes that the lint against the reassignment creation of a `&mut` to a local variable not
marked with `mut` be changed from a forbid-by-default diagnostic to a deny-by-default diagnostic (called
`mut_non_mut` for the sake of discussion).

This will allow users, if they wish, to `allow(mut_non_mut)` or `warn(mut_non_mut)`.
```

Reply:
You're right, but I think the existing language is more precise.  Lints in rustc are defined as configurable diagnostics, so a change from an unconfigurable diagnostic to a configurable one is a change from a non-lint to a lint.  I like that final sentence, though, so I added it in.

Reply:
It's not more precise; it contains a category error. Whether or not a diagnostic is a warning or error is a mix of compiler and user configuration. It's a category error to say "we're turning an error into a lint", since lints can be (and often are) errors. A user `deny`ing or `forbid`'ing a lint does not make the lint cease to be a lint.

Moreover, since RFCs are for user-facing changes, they should use user-facing language. It doesn't matter to a user how the compiler technically implements a diagnostic.

(An [MCP](https://forge.rust-lang.org/compiler/mcp.html), on the other hand, should reference the implementation details of a diagnostic. It'd be totally appropriate to file an MCP proposing that the `E0596` diagnostic be moved from `rustc_borrowck` to `rustc_errors` as a forbid-by-default lint, since that's not a user-facing change.)

Reply:
> It's a category error to say "we're turning an error into a lint", since
> lints can be (and often are) errors

Perhaps my use of the term "hard error" is too ambiguous?  I was using "hard" to
mean "hard-coded"/"non-configurable," but I can see how it might be unclear.

> Moreover, since RFCs are for user-facing changes, they should use user-facing
> language. It doesn't matter to a user how the compiler technically implements
> a diagnostic.

I believe the usage of the term "lint" in user-facing rust documentation is consistent with
the "configurable diagnostic" definition.  Anecdotally, I've seen the term used that
way in various online rust spaces as well.

The actual compiler output admittedly doesn't make a strong distinction.

Reply:
I don't think it is accurate to call the current diagnostic forbid-by-default -- that would imply that it is a lint that has a name that can be set in `allow`/... and that is set to forbid-by-default.

The typical term we use for these is "hard error" (which is of the same category as "lint"), so it is completely in line with standard long-standing compiler terminology to say that something is converted from a hard error to an X-by-default lint.